From ad330cef5e3c157e4564b114aafa192c9c88657d Mon Sep 17 00:00:00 2001
From: Davide Garberi <dade.garberi@gmail.com>
Date: Wed, 21 Aug 2019 03:38:48 +0200
Subject: [PATCH 140/178] leds: Import xiaomi longcheer changes

* Changed an ifdef
* Removed partial lavender changes

Signed-off-by: Isaac Chen <tingyi364@gmail.com>

drivers: leds: Revert xiaomi changes on leds-qpnp

* We want things to be actually working huh

This partially reverts commit feb2e318159c8ad38371c992b9138f38a0f22b71.

Change-Id: Ib5722008533bdca43ff941a419e9bc5f68280a0c
---
 drivers/leds/led-class.c          |   2 +
 drivers/leds/leds-qpnp-flash-v2.c | 200 ++++++++++++++++++++++++++++++
 2 files changed, 202 insertions(+)

diff --git a/drivers/leds/led-class.c b/drivers/leds/led-class.c
index 77b7ea8291c5..e11c25065da1 100644
--- a/drivers/leds/led-class.c
+++ b/drivers/leds/led-class.c
@@ -54,8 +54,10 @@ static ssize_t brightness_store(struct device *dev,
 	if (ret)
 		goto unlock;
 
+#ifndef CONFIG_MACH_XIAOMI_TULIP
 	if (state == LED_OFF && !(led_cdev->flags & LED_KEEP_TRIGGER))
 		led_trigger_remove(led_cdev);
+#endif
 	led_set_brightness(led_cdev, state);
 	led_cdev->usr_brightness_req = state;
 
diff --git a/drivers/leds/leds-qpnp-flash-v2.c b/drivers/leds/leds-qpnp-flash-v2.c
index d00ce6d5108c..602bc0781856 100644
--- a/drivers/leds/leds-qpnp-flash-v2.c
+++ b/drivers/leds/leds-qpnp-flash-v2.c
@@ -277,6 +277,21 @@ struct flash_switch_data {
 	bool				symmetry_en;
 };
 
+#ifdef CONFIG_MACH_LONGCHEER
+struct flashlight_node_data {
+	struct platform_device		*pdev;
+	const char			**torch_name;
+	const char			**switch_name;
+	u8				num_torch;
+	u8				num_switch;
+	struct led_classdev		cdev;
+	u8				id;
+	bool				led_on;
+};
+
+#define NAME_SIZE	20
+#endif
+
 /*
  * Flash LED configuration read from device tree
  */
@@ -331,6 +346,9 @@ struct qpnp_flash_led {
 	struct regmap			*regmap;
 	struct flash_node_data		*fnode;
 	struct flash_switch_data	*snode;
+#ifdef CONFIG_MACH_LONGCHEER
+	struct flashlight_node_data	*flashlight_node;
+#endif
 	struct power_supply		*bms_psy;
 	struct power_supply		*main_psy;
 	struct power_supply		*usb_psy;
@@ -338,6 +356,9 @@ struct qpnp_flash_led {
 	spinlock_t			lock;
 	int				num_fnodes;
 	int				num_snodes;
+#ifdef CONFIG_MACH_LONGCHEER
+	int				num_flashlight_nodes;
+#endif
 	int				enable;
 	int				total_current_ma;
 	u32				wa_flags;
@@ -1936,8 +1957,75 @@ static void qpnp_flash_led_brightness_set(struct led_classdev *led_cdev,
 		qpnp_flash_led_node_set(fnode, value);
 	}
 
+#ifdef CONFIG_MACH_LONGCHEER
+	if (!strcmp("led:switch_0", led_cdev->name) && !value)
+		if (led->flashlight_node != NULL)
+			led->flashlight_node->cdev.brightness = value;
+#endif
+
+	spin_unlock(&led->lock);
+}
+
+#ifdef CONFIG_MACH_LONGCHEER
+static void qpnp_flashlight_led_brightness_set(struct led_classdev *led_cdev,
+						enum led_brightness value)
+{
+	struct flashlight_node_data *flashlight_data = NULL;
+	struct qpnp_flash_led *led = NULL;
+	int rc;
+	int i, j;
+
+#ifdef CONFIG_MACH_XIAOMI_LAVENDER
+	if (100 == value)
+		value = 70;
+#endif
+
+	if (!strcmp("flashlight", led_cdev->name)) {
+		flashlight_data = container_of(led_cdev, struct flashlight_node_data, cdev);
+		led = dev_get_drvdata(&flashlight_data->pdev->dev);
+	}
+
+	if (!led) {
+		pr_err("Failed to get flash driver data\n");
+		return;
+	}
+
+	spin_lock(&led->lock);
+	if (flashlight_data) {
+		for(i = 0; i < flashlight_data->num_switch; ++i)
+			for(j = 0; j < led->num_snodes; ++j) {
+				pr_debug(" switch name[%d] = %s, snode name[%d] = %s\n",i, flashlight_data->switch_name[i],j, led->snode[j].cdev.name);
+				if (!strcmp(flashlight_data->switch_name[i], led->snode[j].cdev.name)) {
+					rc = qpnp_flash_led_switch_set(&led->snode[j], false);
+					if (rc < 0)
+						pr_err("Failed to set flash LED switch rc=%d\n", rc);
+					break;
+				}
+			}
+
+		for(i = 0; i < flashlight_data->num_torch; ++i)
+			for(j = 0; j < led->num_fnodes; ++j) {
+				pr_debug(" torch name[%d] = %s, fnode name[%d] = %s\n",i, flashlight_data->torch_name[i],j, led->fnode[j].cdev.name);
+				if (!strcmp(flashlight_data->torch_name[i], led->fnode[j].cdev.name)) {
+					qpnp_flash_led_node_set(&led->fnode[j], value);
+					break;
+				}
+			}
+
+		for(i = 0; i < flashlight_data->num_switch; ++i)
+			for(j = 0; j < led->num_snodes; ++j) {
+				pr_debug(" switch name[%d] = %s, snode name[%d] = %s\n",i, flashlight_data->switch_name[i],j, led->snode[j].cdev.name);
+				if (!strcmp(flashlight_data->switch_name[i], led->snode[j].cdev.name)) {
+					rc = qpnp_flash_led_switch_set(&led->snode[j], value > 0);
+					if (rc < 0)
+						pr_err("Failed to set flash LED switch rc=%d\n", rc);
+					break;
+				}
+			}
+	}
 	spin_unlock(&led->lock);
 }
+#endif
 
 static ssize_t qpnp_flash_led_prepare_store(struct device *dev,
 		struct device_attribute *attr, const char *buf, size_t count)
@@ -2315,6 +2403,11 @@ static int qpnp_flash_led_parse_each_led_dt(struct qpnp_flash_led *led,
 	if (rc < 0)
 		return rc;
 
+#ifdef CONFIG_MACH_XIAOMI_LAVENDER
+	if (fnode->type == FLASH_LED_TYPE_TORCH)
+		fnode->cdev.flags |= LED_KEEP_TRIGGER;
+#endif
+
 	rc = led_classdev_register(&led->pdev->dev, &fnode->cdev);
 	if (rc < 0) {
 		pr_err("Unable to register led node %d\n", fnode->id);
@@ -2335,6 +2428,83 @@ static int qpnp_flash_led_parse_each_led_dt(struct qpnp_flash_led *led,
 	return 0;
 }
 
+#ifdef CONFIG_MACH_LONGCHEER
+static int qpnp_flashlight_led_parse_and_register(struct qpnp_flash_led *led,
+			struct flashlight_node_data *flashlight_node, struct device_node *node)
+{
+	int rc,i;
+	u32 val,count;
+	const char **temp_name;
+
+	flashlight_node->pdev = led->pdev;
+	flashlight_node->cdev.brightness_set = qpnp_flashlight_led_brightness_set;
+	flashlight_node->cdev.brightness_get = qpnp_flash_led_brightness_get;
+
+	rc = of_property_read_string(node, "qcom,led-name", &flashlight_node->cdev.name);
+	if (rc < 0) {
+		pr_err("Unable to read flash LED names\n");
+		return rc;
+	}
+
+	count = of_property_count_strings(node, "qcom,torch-name");
+	if (!count || (count == -EINVAL)) {
+		pr_err("%s:%d number of entries is 0 or not present in dts\n",
+			__func__, __LINE__);
+		return -EINVAL;
+	} else
+		flashlight_node->num_torch = count;
+	pr_debug("%s qcom,torch-name count %d\n", __func__, flashlight_node->num_torch);
+
+	temp_name = kzalloc(sizeof(char *) * count, GFP_KERNEL);
+	if (temp_name) {
+		for(i = 0; i < count; ++i) {
+			flashlight_node->torch_name = temp_name;
+			temp_name[i] = kzalloc(sizeof(char) * NAME_SIZE,GFP_KERNEL);
+			rc = of_property_read_string_index(node,"qcom,torch-name", i,&temp_name[i]);
+			pr_debug("%s torch_name[%d] = %s\n", __func__, i,flashlight_node->torch_name[i]);
+		}
+	} else
+		pr_err("%s alloc torch_name failed!!!\n",__func__);
+
+	count = of_property_count_strings(node, "qcom,switch-name");
+	if (!count || (count == -EINVAL)) {
+		pr_err("%s:%d number of entries is 0 or not present in dts\n",
+			__func__, __LINE__);
+		return -EINVAL;
+	} else
+		flashlight_node->num_switch = count;
+	pr_debug("%s qcom,switch-name count %d\n", __func__, count);
+
+	temp_name = kzalloc(sizeof(char *) * count, GFP_KERNEL);
+	if (temp_name) {
+		for(i = 0; i < count; ++i) {
+			flashlight_node->switch_name = temp_name;
+			temp_name[i] = kzalloc(sizeof(char) * NAME_SIZE,GFP_KERNEL);
+			rc = of_property_read_string_index(node,"qcom,switch-name", i,&temp_name[i]);
+			pr_debug("%s switch_name[%d] = %s\n", __func__, i,flashlight_node->switch_name[i]);
+		}
+	} else
+		pr_err("%s alloc switch_name faild!!!\n",__func__);
+
+	rc = of_property_read_u32(node, "qcom,id", &val);
+	if (!rc) {
+		flashlight_node->id = (u8)val;
+	} else {
+		pr_err("Unable to read flashlight LED ID\n");
+		return rc;
+	}
+
+	rc = led_classdev_register(&led->pdev->dev, &flashlight_node->cdev);
+	if (rc < 0) {
+		pr_err("Unable to register flashlight node %d\n", flashlight_node->id);
+		return rc;
+	}
+
+	flashlight_node->cdev.dev->of_node = node;
+	return 0;
+}
+#endif
+
 static int qpnp_flash_led_parse_and_register_switch(struct qpnp_flash_led *led,
 						struct flash_switch_data *snode,
 						struct device_node *node)
@@ -3094,6 +3264,9 @@ static int qpnp_flash_led_probe(struct platform_device *pdev)
 	struct device_node *node, *temp;
 	const char *temp_string;
 	int rc, i = 0, j = 0;
+#ifdef CONFIG_MACH_LONGCHEER
+	int k = 0;
+#endif
 
 	node = pdev->dev.of_node;
 	if (!node) {
@@ -3139,6 +3312,10 @@ static int qpnp_flash_led_probe(struct platform_device *pdev)
 		} else if (!strcmp("flash", temp_string) ||
 				!strcmp("torch", temp_string)) {
 			led->num_fnodes++;
+#ifdef CONFIG_MACH_LONGCHEER
+		} else if (!strcmp("flashlight", temp_string)) {
+			led->num_flashlight_nodes++;
+#endif
 		} else {
 			pr_err("Invalid label for led node\n");
 			return -EINVAL;
@@ -3160,9 +3337,20 @@ static int qpnp_flash_led_probe(struct platform_device *pdev)
 	if (!led->snode)
 		return -ENOMEM;
 
+#ifdef CONFIG_MACH_LONGCHEER
+	led->flashlight_node = devm_kcalloc(&pdev->dev, led->num_flashlight_nodes,
+				sizeof(*led->flashlight_node),
+				GFP_KERNEL);
+	if (!led->flashlight_node)
+		return -ENOMEM;
+#endif
+
 	temp = NULL;
 	i = 0;
 	j = 0;
+#ifdef CONFIG_MACH_LONGCHEER
+	k = 0;
+#endif
 	for_each_available_child_of_node(node, temp) {
 		rc = of_property_read_string(temp, "label", &temp_string);
 		if (rc < 0) {
@@ -3192,6 +3380,18 @@ static int qpnp_flash_led_probe(struct platform_device *pdev)
 			}
 			j++;
 		}
+#ifdef CONFIG_MACH_LONGCHEER
+		if (!strcmp("flashlight", temp_string)) {
+			rc = qpnp_flashlight_led_parse_and_register(led,
+					&led->flashlight_node[k],temp);
+			if (rc < 0) {
+				pr_err("Unable to parse and register flashlight node, rc=%d\n",
+					rc);
+				goto error_switch_register;
+			}
+			k++;
+		}
+#endif
 	}
 
 	rc = qpnp_flash_led_register_interrupts(led);
-- 
2.49.0

